cpp/ql/test/library-tests/dataflow/dataflow-tests/ref.cpp

```diff
@@ -0,0 +1,134 @@
+int source();
+
+template<typename T>
+void sink(T);
+
+extern int arbitrary;
+
+namespace withoutFields {
+  template<typename T>
+  void assign(T &lhs, T rhs) {
+    lhs = rhs;
+  }
+
+  template<typename T>
+  void assignWrapper(T &lhs, T rhs) {
+    assign(lhs, rhs);
+  }
+
+  void notAssign(int &lhs, int rhs) {
+    lhs = rhs;
+    if (arbitrary) {
+      lhs = 1;
+    } else {
+      lhs = 2;
+    }
+  }
+
+  void sourceToParam(int &out) {
+    out = source();
+    if (arbitrary) {
+      out = 1;
+    }
+  }
+
+  void sourceToParamWrapper(int &out) {
+    if (arbitrary) {
+      sourceToParam(out);
+    } else {
+      out = 1;
+    }
+  }
+
+  void notSource(int &out) {
+    out = source();
+    if (arbitrary) {
+      out = 1;
+    } else {
+      out = 2;
+    }
+  }
+
+  void testRefs() {
+    int x1, x2, x3, x4;
+
+    assignWrapper(x1, source());
+    sink(x1); // flow [FALSE POSITIVE from uninitialized]
+
+    notAssign(x2, source());
+    sink(x2); // no flow [FALSE POSITIVE from uninitialized]
+
+    sourceToParamWrapper(x3);
+    sink(x3); // flow [FALSE POSITIVE from uninitialized]
+
+    notSource(x4);
+    sink(x4); // no flow [FALSE POSITIVE from uninitialized]
```

Comment:
What's the story behind these false positives?

Reply:
The data flow library has a rarely-used feature for detecting uninitialized data as a source. The tests in this directory configure all uninitialized variables to be sources in order to exercise that feature. In practice, too many things are considered to be uninitialized. This is especially the case around definition-by-reference. For example:

```
mystruct x;
assign_to_pointer(&x);
```

Is `x` uninitialized after this? Probably not. But to be sure, we'd need to look carefully at the body of `assign_to_pointer` and possibly at the functions it calls. What if some but not all fields of `x` are assigned to? What if the padding between fields is unassigned? What if there's a rare error condition that causes `x` to be left unassigned?

Without adding a very expensive analysis of what's overwritten, it's hard to pick a good default here. See #1186 and CPP-336 for historical context.

Reply:
I'm happy with your explanation.  My intuition would also be that `call(&x)` should be assumed to initialize `x` (assuming it isn't a `const` reference parameter) to avoid false positives for queries about uninitialized variables / data.  We should keep an eye on these kinds of situations in the real world.